java/org/apache/tomcat/websocket/WsSession.java

```diff
@@ -816,16 +823,62 @@ protected void checkExpiration() {
             return;
         }
 
-        if (System.currentTimeMillis() - lastActive > timeout) {
+        long currentTime = System.currentTimeMillis();
+
+        if (currentTime - lastActive > timeout) {
             String msg = sm.getString("wsSession.timeout", getId());
             if (log.isDebugEnabled()) {
                 log.debug(msg);
             }
             doClose(new CloseReason(CloseCodes.GOING_AWAY, msg),
                     new CloseReason(CloseCodes.CLOSED_ABNORMALLY, msg));
+        } else {
+            boolean closeConnection = false;
+            Long idleReadTimeout = getIdleReadTimeout();
+            
+            closeConnection = idleReadTimeout != null && (currentTime - lastReadTime > idleReadTimeout);
+
+            if (closeConnection) {
+                String msg = sm.getString("wsSession.readIdleTimeout", getId());
+                if (log.isDebugEnabled()) {
+                    log.debug(msg);
+                }
+                doClose(new CloseReason(CloseCodes.GOING_AWAY, msg),
+                        new CloseReason(CloseCodes.CLOSED_ABNORMALLY, msg));
+            }
+
+            Long idleWriteTimeout = getIdleWriteTimeout();
+            
+            closeConnection = idleWriteTimeout != null && (currentTime - lastWriteTime > idleWriteTimeout);
+
+            if (closeConnection) {
+                String msg = sm.getString("wsSession.writeIdleTimeout", getId());
+                if (log.isDebugEnabled()) {
+                    log.debug(msg);
+                }
+                doClose(new CloseReason(CloseCodes.GOING_AWAY, msg),
+                        new CloseReason(CloseCodes.CLOSED_ABNORMALLY, msg));
+            }
```

Comment:
Don't need to perform this check if the read timeout has already triggered. Don't want the possibility of multiple close messages.